lib/oli/delivery/attempts/activity_lifecycle/roll_up.ex

```diff
@@ -0,0 +1,542 @@
+    - `part_attempts`: The current part attempts.
+  - **Behavior**: Based on the inputs and attempt states, selects one of:
+    - `evaluated_fn`: Rolls up to activity and possibly resource level.
+    - `submitted_fn`: Marks activity as submitted without scoring rollup.
+    - `no_op_fn`: Does nothing (noop).
+  - **Returns**: A zero-arity function to be executed later.
+
+  ---
+
+  #### `rollup_evaluated/2`
+  Performs a full rollup of evaluated part attempts to their corresponding activity attempt and,
+  if applicable, to the resource attempt and resource access records.
+
+  - **Params**:
+    - `activity_attempt_guid`: The GUID of the evaluated activity attempt.
+    - `part_attempts`: List of part attempts used for score aggregation.
+  - **Behavior**:
+    - Calculates aggregate scores based on the configured scoring strategy.
+    - Updates `activity_attempt`, `resource_attempt`, and `resource_access` records.
+    - Supports both "batch scoring" and "score-as-you-go" workflows.
+  - **Returns**: `:ok` or `:error`.
+
+  ---
+
+  #### `rollup_submitted/1`
+  Marks an activity attempt as `:submitted` without performing score aggregation.
+  """
+
+  require Logger
+
+  import Ecto.Query, warn: false
+  import Oli.Delivery.Attempts.Core
+  import Oli.Delivery.Attempts.ActivityLifecycle.Persistence
+  import Oli.Delivery.Attempts.ActivityLifecycle.Utils
+
+  alias Oli.Delivery.Attempts.Core.{ActivityAttempt, ResourceAttempt, ResourceAccess}
+  alias Oli.Delivery.Evaluation.{Result}
+  alias Oli.Delivery.Attempts.Core.StudentInput
+  alias Oli.Repo
+  alias Oli.Delivery.Attempts.Scoring
+  alias Oli.Resources.Revision
+  alias Oli.Delivery.Sections.Section
+  alias Oli.Delivery.Sections.SectionResource
+
+  @doc """
+  Returns a function that when executed, will properly roll up the evaluated
+  part attempts to (at least) the activity attempt. Depending on the context,
+  this function may also roll up the results to the resource attempt and resource access.
+  """
+  def determine_activity_rollup_fn(
+        activity_attempt_guid,
+        part_inputs,
+        part_attempts
+      ) do
+    evaluated_fn = fn result ->
+      rollup_evaluated(activity_attempt_guid)
+      result
+    end
+
+    submitted_fn = fn result ->
+      rollup_submitted(activity_attempt_guid)
+      result
+    end
+
+    no_op_fn = fn result -> result end
+
+    case determine_activity_rollup_state(part_inputs, part_attempts) do
+      :evaluated ->
+        evaluated_fn
+
+      :submitted ->
+        submitted_fn
+
+      :no_op ->
+        no_op_fn
+    end
+  end
+
+  @doc """
+  From the update to a collection of part attempts, return a function that will properly roll
+  up the score to the activity attempt, and possibly the resource attempt and resource access.
+
+  The returned function takes zero arguments, and returns either :ok or :error.
+
+  This function makes at most two read queries to the database.
```

Comment:
Nit: Function docs say this function makes at most 2 queries, but it looks if it is score as you go it will
make 3